ibm/service/power/resource_ibm_pi_virtual_serial_number.go

```diff
@@ -0,0 +1,398 @@
+
+		} else {
+			serialString = vsnArg
+		}
+	}
+
+	id := cloudInstanceID + "/" + serialString
+	d.SetId(id)
+
+	return resourceIBMPIVirtualSerialNumberRead(ctx, d, meta)
+}
+
+func resourceIBMPIVirtualSerialNumberRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	idArr, err := flex.IdParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := idArr[0]
+	serial := idArr[1]
+	client := instance.NewIBMPIVSNClient(ctx, sess, cloudInstanceID)
+
+	vsn, err := client.Get(serial)
+	if err != nil {
+		if strings.Contains(strings.ToLower(err.Error()), NotFound) {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+	d.Set(Arg_Description, vsn.Description)
+	d.Set(Arg_InstanceID, vsn.PvmInstanceID)
+	d.Set(Arg_Serial, vsn.Serial)
+
+	return nil
+}
+
+func resourceIBMPIVirtualSerialNumberDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	idArr, err := flex.IdParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	cloudInstanceID := idArr[0]
+	client := instance.NewIBMPIVSNClient(ctx, sess, cloudInstanceID)
+
+	if v, ok := d.GetOk(Arg_InstanceID); ok {
+		pvmInstanceId := v.(string)
+		instanceClient := instance.NewIBMPIInstanceClient(ctx, sess, cloudInstanceID)
+		restartInstance, err := stopLparForVSNChange(ctx, instanceClient, pvmInstanceId, d.Timeout(schema.TimeoutDelete))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		retainVSN := false
+		if v, ok := d.GetOk(Arg_RetainVirtualSerialNumber); ok {
+			retainVSN = v.(bool)
+		}
+		deleteBody := &models.DeleteServerVirtualSerialNumber{
+			RetainVSN: retainVSN,
+		}
+		err = client.PVMInstanceDeleteVSN(pvmInstanceId, deleteBody)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		_, err = isWaitForPIInstanceStopped(ctx, instanceClient, pvmInstanceId, d.Timeout(schema.TimeoutUpdate))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		if restartInstance {
+			err = startLparAfterVSNChange(ctx, instanceClient, pvmInstanceId, d.Timeout(schema.TimeoutDelete))
+			if err != nil {
+				return diag.FromErr(err)
+			}
+		}
+
+	} else {
+		serialNumber := d.Get(Arg_Serial).(string)
+		err = client.Delete(serialNumber)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func resourceIBMPIVirtualSerialNumberUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	cloudInstanceID := d.Get(Arg_CloudInstanceID).(string)
+	client := instance.NewIBMPIVSNClient(ctx, sess, cloudInstanceID)
+
+	if d.HasChange(Arg_Description) && !d.HasChange(Arg_InstanceID) {
```

Comment:
what happens if you have a change in the description and the instance?

Reply:
It will go to the next if statement and the description will get added to the request for during the instance change request:

```
                        serial := d.Get(Arg_Serial).(string)
			addBody := &models.AddServerVirtualSerialNumber{
				Serial: &serial,
			}
			if v, ok := d.GetOk(Arg_Description); ok {
				description := v.(string)
				addBody.Description = description
			}
```